src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -259,12 +258,64 @@ uct_rdmacm_cm_ep_create_reserved_qpn(uct_rdmacm_cm_ep_t *cep,
         ucs_list_add_tail(&ctx->blk_list, &blk->entry);
     }
 
-    cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset;
+    if (ctx->reuse_qpn) {
+        status = uct_rdmacm_cm_get_peer_dev_ctx(ctx, &cep->id->route,
+                                                &peer_dev_ctx);
+        if (status != UCS_OK) {
+            goto out;
+        }
+
+        if (peer_dev_ctx->ref_qpn_blk == NULL) {
+            blk                                 = ucs_list_next(&ctx->blk_list,
+                                                                uct_rdmacm_cm_reserved_qpn_blk_t, entry);
+            peer_dev_ctx->ref_qpn_blk           = blk;
+            peer_dev_ctx->next_avail_qpn_offset = 0;
+        } else if (peer_dev_ctx->next_avail_qpn_offset >= qpns_per_obj) {
+            if (peer_dev_ctx->ref_qpn_blk->entry.next == &ctx->blk_list) {
+                status = uct_rdmacm_cm_reserved_qpn_blk_alloc(
+                        ctx, cep->id->verbs, UCS_LOG_LEVEL_ERROR, &blk);
+                if (status != UCS_OK) {
+                    goto out;
+                }
+
+                ucs_list_add_tail(&ctx->blk_list, &blk->entry);
+            }
+            blk                                 = ucs_list_next(&peer_dev_ctx->ref_qpn_blk->entry,
+                                                                uct_rdmacm_cm_reserved_qpn_blk_t, entry);
+            peer_dev_ctx->ref_qpn_blk           = blk;
+            peer_dev_ctx->next_avail_qpn_offset = 0;
+        } else {
+            blk = peer_dev_ctx->ref_qpn_blk;
+        }
+    } else {
+        blk = ucs_list_tail(&ctx->blk_list, uct_rdmacm_cm_reserved_qpn_blk_t,
+                            entry);
+        if (blk->next_avail_qpn_offset >= qpns_per_obj) {
+            status = uct_rdmacm_cm_reserved_qpn_blk_alloc(ctx, cep->id->verbs,
+                                                          UCS_LOG_LEVEL_ERROR,
+                                                          &blk);
+            if (status != UCS_OK) {
+                goto out;
+            }
+
+            ucs_list_add_tail(&ctx->blk_list, &blk->entry);
+        }
+    }
+
+    if (ctx->reuse_qpn) {
+        ucs_assert(peer_dev_ctx != NULL);
+        cep->qpn                   = blk->first_qpn +
+                                     peer_dev_ctx->next_avail_qpn_offset++;
+        blk->next_avail_qpn_offset = ucs_max(blk->next_avail_qpn_offset,
+                                             peer_dev_ctx->next_avail_qpn_offset);
+    } else {
+        cep->qpn                   = blk->first_qpn +
+                                     blk->next_avail_qpn_offset++;
+    }
```

Comment:
@evgeny-leksikov 
1. In the existed code , ``cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset++```. So we both agree with this code.
2. We need to understand the case ```if (ctx->reuse_qpn) {}```, because this is the new feature. Let me explain here:
    1) According to the design document,  ```cep->qpn = peer_dev_ctx->ref_qpn_blk->first_qpn + peer_dev_ctx->next_avail_qpn_offset++```. Since ```blk = peer_dev_ctx->ref_qpn_blk``` when ```REUSE_QPN=ON```, so the code could be implemented as ```cep->qpn = blk->first_qpn + peer_dev_ctx->next_avail_qpn_offset++```.
    2) For the logic ```blk->next_avail_qpn_offset = ucx_max(blk->next_avail_qpn_offset, peer_dev_ctx->next_avail_qpn_offset);```, this is needed for releasing the reserved_qpn and its corresponding block. Please check the existed code: https://github.com/openucx/ucx/blob/v1.14.0/src/uct/ib/rdmacm/rdmacm_cm_ep.c#L296

Let's talk about this piece of code in the PR as we've talked in teams.